api/v1alpha3/docluster_webhook.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var _ = logf.Log.WithName("docluster-resource")
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-docluster,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=doclusters,versions=v1alpha3,name=validation.docluster.infrastructure.cluster.x-k8s.io,sideEffects=None
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha3-docluster,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=doclusters,versions=v1alpha3,name=default.docluster.infrastructure.cluster.x-k8s.io,sideEffects=None
+
+var (
+	_ webhook.Defaulter = &DOCluster{}
+	_ webhook.Validator = &DOCluster{}
+)
+
+func (r *DOCluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *DOCluster) Default() {}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *DOCluster) ValidateCreate() error {
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *DOCluster) ValidateUpdate(old runtime.Object) error {
+	var allErrs field.ErrorList
+
+	oldDOCluster, ok := old.(*DOCluster)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an DOCluster but got a %T", old))
+	}
+
+	if r.Spec.Region != oldDOCluster.Spec.Region {
```

Comment:
The one case I can think of where changing the region _might_ be helpful is when a wrong region was specified in the first place and you were trying to fix a reconcile that fails early on. Might not be worth considering though.

Reply:
make sense. but I prefer to make the region immutable in some cases the user trying to change `region` in an existing cluster and then the next machine will be created in a different region from the other old machines.

how about make validation in the `ValidateCreate` that region must be valid do region? so it will reduce the failure because of the invalid region.

Reply:
I like the idea to check if the region is a valid one. We should then use [the regions endpoint to list all valid regions](https://developers.digitalocean.com/documentation/v2/#list-all-regions).

(Doesn't have to implemented in this PR.)